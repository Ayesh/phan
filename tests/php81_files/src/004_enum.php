<?php
namespace Enums {
}

enum X: int {
    case Foo = 2;
    case Bar = 3;
    case Baz = 4;
    const Baz = 4; // should warn
}
var_export(X::Foo === 2);
X::Bar();  // should warn about missing method
(X::Bar)();  // should warn about missing __invoke
class Invalid extends X {}

enum HasDuplicate: string {
    case IS_EMPTY      = '';
    case IS_ZERO       = '0'; // duplicates are fatal errors
    case IS_ALSO_ZERO  = '0';
    case IS_ALSO_EMPTY = '' . '';
    case Invalid = $var;
}
new HasDuplicate(); // this is an error

enum InvalidEnum {
    case Valid;
    case Foo = null;

    abstract function mustNotBeAbstract();
}
