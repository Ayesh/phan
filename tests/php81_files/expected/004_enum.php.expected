%s:8 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\X::Baz
%s:9 PhanRedefineClassConstant Class constant Baz defined at %s:9 was previously defined at %s:8
%s:11 PhanImpossibleTypeComparison Impossible attempt to check if X::Foo of type \Enums\X is identical to 2 of type 2
%s:12 PhanUndeclaredStaticMethod Static call to undeclared method \Enums\X::Bar
%s:13 PhanTypeInvalidCallable Saw type \Enums\X which cannot be a callable
%s:13 PhanUndeclaredInvokeInCallable Possible attempt to access missing magic method __invoke of '\Enums\X'
%s:14 PhanAccessExtendsFinalClass Attempting to extend from final class \Enums\X defined at %s:5
%s:14 PhanUnreferencedClass Possibly zero references to class \Enums\Invalid
%s:17 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\HasDuplicate::IS_EMPTY
%s:18 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\HasDuplicate::IS_ZERO
%s:19 PhanReusedEnumCaseValue Enum case IS_ALSO_ZERO has the same value('0') as a previous declared enum case IS_ZERO defined at %s:18
%s:19 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\HasDuplicate::IS_ALSO_ZERO
%s:20 PhanReusedEnumCaseValue Enum case IS_ALSO_EMPTY has the same value('') as a previous declared enum case IS_EMPTY defined at %s:17
%s:20 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\HasDuplicate::IS_ALSO_EMPTY
%s:21 PhanInvalidConstantExpression Constant expression contains invalid operations ($var)
%s:21 PhanUndeclaredGlobalVariable Global variable $var is undeclared
%s:21 PhanUndeclaredVariable Variable $var is undeclared
%s:21 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\HasDuplicate::Invalid
%s:23 PhanNoopNewNoSideEffects Unused result of new object creation expression in new HasDuplicate() (this is likely free of side effects - there is no known non-empty constructor or destructor)
%s:23 PhanTypeInstantiateEnum Saw instantiation of enum \Enums\HasDuplicate
%s:25 PhanClassContainsAbstractMethod non-abstract class \Enums\InvalidEnum contains abstract method \Enums\InvalidEnum::mustNotBeAbstract() declared at %s:29
%s:25 PhanUnreferencedClass Possibly zero references to enum \Enums\InvalidEnum
%s:26 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\InvalidEnum::Valid
%s:27 PhanSyntaxEnumCaseUnexpectedValue Syntax error: Expected enum case Foo not to have a value
%s:27 PhanUnreferencedPublicClassConstant Possibly zero references to public class constant \Enums\InvalidEnum::Foo
